### PR TITLE
docs: document all unstable configurations (#20336)

### LINKED ISSUE #20336: Improve documentation on experimetal (alpha/beta) features
# Summary

When a new feature is added as an alpha or beta, it often requires changes to the ArgoCD CRD, either in the spec, the status or both. Apart from the documentation on the specific feature and admonition warnings, it is quite hard for a user to distinguish what is GA/Stable from what is not.

This tends to create the false belief that anything in the spec will not be removed or have breaking changes, which is not true based on https://github.com/argoproj/argoproj/blob/b1f9b68795fe75faa3ef38f31696fad6c6b3ebf7/community/feature-status.md.

# Motivation

The goal is never to purposely break the user workflows by introducing breaking changes or removing features. However, we rarely get it right the first time! 

Sometime using annotations can be a good alternative to avoid modifying the spec. However, it has at least 3 major downside:
1. Needs to be re-implement to spec/status to progress to beta or stable, which affect user experience.
2. Some existing annotations are already considered stable.
3. 

I think modifying the spec/status is necessary to implement new features and perform experimentation as first class citizens.

# Proposal

To reduce the possible user confusion while maintaining a good user experience for adopters of alpha and beta features, I think we should do the following:

1. Clarify the documentation
  a. Create a new documentation page with all the current alpha and beta features, as well as the explicit list of fields in the spec affected.
  b. Duplicate the information feature status information from [argoproj/argoproj](https://github.com/argoproj/argoproj/blob/b1f9b68795fe75faa3ef38f31696fad6c6b3ebf7/community/feature-status.md.) in argo-cd docs because most people will not read that documentation.
2. Add CRD documentation on the fields that are alpha/beta. If possible, it would be nice to have these fields lint warnings in IDEs when they are used.

With better documentation, we will increase user awareness when they decide to adopt features that are under development

## Alternative (long-term)

I think we should start considering using the Kubernetes apiVersion, which is currently v1alpha1, to allow us to perform these changes in a more Kubernetes way. I don't have much experience on this, but definitely think we should start implementing a versioning strategy + mutation webhook to manage the apiVersions.

Some examples: 
- https://gateway-api.sigs.k8s.io/concepts/versioning/#release-channels-eg-experimental-standard (thanks @zachaller)
- https://book.kubebuilder.io/multiversion-tutorial/api-changes

### PR DESCRIPTION
**Pull request by agaudreault**
Closes #20336

Add all the fields in the doc that are subject to breaking changes or removal. This aims to clarify what should and should not be used by users and/or vendors.

Previous discussions with the community highlighted that spec fields on the CRD are falsely considered more "stable", than annotations.

### CODE REVIEW

File: docs/operator-manual/feature-maturity.md [hunk 1]
```diff
@@ -3,22 +3,83 @@
...
 [4]: ../user-guide/skip_reconcile.md
 [5]: applicationset/Appset-Any-Namespace.md
-[6]: dynamic-cluster-distribution.md
-[7]: ../user-guide/diff-strategies.md#server-side-diff
-[8]: app-sync-using-impersonation.md
+[6]: ./high_availability.md#argocd-application-controller
+[7]: dynamic-cluster-distribution.md
+[8]: ../user-guide/diff-strategies.md#server-side-diff
+[9]: ./high_availability.md#argocd-application-controller
+[10]: app-sync-using-impersonation.md
```
Comment (leoluz) [inline docs/operator-manual/feature-maturity.md:37]: Please always keep the link section at the bottom of the file for easy reference.

**Review by leoluz — CHANGES_REQUESTED** [review]

File: docs/operator-manual/feature-maturity.md (hunk 1 above)
Comment (agaudreault) [reply docs/operator-manual/feature-maturity.md:37]: The links section can be placed anywhere in the file. In this case, I think it is much easier to have these link directly under the main overview table.

Comment (leoluz) [reply docs/operator-manual/feature-maturity.md:37]: You are referencing the links in the new section below which makes it confusing to find where they are defined. This type of markdown links are called Reference and usually you put references at the bottom of the file. You can read more about it here:
https://google.github.io/styleguide/docguide/style.html#reference

**Review by leoluz — CHANGES_REQUESTED** [review]

File: docs/operator-manual/feature-maturity.md (hunk 1 above)
Comment (agaudreault) [reply docs/operator-manual/feature-maturity.md:37]: Let me explain in more details why I think, in this case, reference links should be put under the main table: if a contributor updates the page to add their new fields to the list, it is very easy to scroll down, add the links and be done with it. When that happens, the main overview table would become incomplete. If the links are placed directly underneath the main table, it becomes more obvious to update the overview table. 

But I'll move them to the bottom if markdown styling is more important.

**Review by leoluz — APPROVED** [review]
LGTM

**Pull request merged**